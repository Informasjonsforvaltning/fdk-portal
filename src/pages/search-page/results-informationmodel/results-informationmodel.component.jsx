import _ from 'lodash';
import React from 'react';
import { withRouter } from 'react-router';
import ReactPaginate from 'react-paginate';
import PropTypes from 'prop-types';
import { Button, Modal, ModalBody, ModalFooter, ModalHeader } from 'reactstrap';
import cx from 'classnames';

import localization from '../../../lib/localization';
import { SearchHitItem } from './search-hit-item/search-hit-item.component';
import { FilterTree } from '../filter-tree/filter-tree.component';
import { getSortfield, setPage, setSortfield } from '../search-location-helper';
import { parseSearchParams } from '../../../lib/location-history-helper';
import { FilterPills } from '../filter-pills/filter-pills.component';

const renderFilterModal = ({
  showFilterModal,
  closeFilterModal,
  locationSearch,
  publisherCounts,
  publishers,
  onFilterPublisherHierarchy
}) => (
  <Modal isOpen={showFilterModal} toggle={closeFilterModal}>
    <ModalHeader toggle={closeFilterModal}>{localization.filter}</ModalHeader>
    <ModalBody>
      <div className="search-filters">
        <FilterTree
          title={localization.responsible}
          aggregations={publisherCounts}
          handleFiltering={onFilterPublisherHierarchy}
          activeFilter={locationSearch.orgPath}
          referenceDataItems={publishers}
        />
      </div>
    </ModalBody>
    <ModalFooter>
      <Button className="fdk-button" onClick={closeFilterModal} color="primary">
        {localization.close}
      </Button>
    </ModalFooter>
  </Modal>
);

const renderHits = (hits, publishers) => {
  if (hits && Array.isArray(hits)) {
    return hits.map((item, index) => (
      <SearchHitItem
        key={item.id}
        item={item}
        fadeInCounter={index < 3 ? index : null}
        publishers={publishers}
      />
    ));
  }
  return null;
};

export const ResultsInformationModelPure = ({
  showFilterModal,
  closeFilterModal,
  informationModelItems,
  informationModelTotal,
  informationModelAggregations,
  onFilterPublisherHierarchy,
  publisherCounts,
  publishers,
  hitsPerPage,
  history,
  location
}) => {
  const locationSearch = parseSearchParams(location);
  const page = parseInt(locationSearch.page || 0, 10);
  const pageCount = Math.ceil((informationModelTotal || 1) / hitsPerPage);

  const sortfield = getSortfield(location);
  const sortByScoreClass = cx('fdk-button', 'fdk-button-black-toggle', {
    selected: !sortfield
  });
  const sortByLastModifiedClass = cx('fdk-button', 'fdk-button-black-toggle', {
    selected: sortfield === 'modified'
  });

  const onSortByScoreClick = () => {
    setSortfield(history, location, undefined);
  };
  const onSortByModifiedClick = () => {
    setSortfield(history, location, 'modified');
  };
  const onPageChange = data => {
    setPage(history, location, data.selected);
    window.scrollTo(0, 0);
  };

  return (
    <main data-test-id="informationModels" id="content">
      <div className="row mb-3">
        <div className="col-6 col-lg-4" />
        <div className="col-6 col-lg-4 offset-lg-4">
          <div className="d-flex justify-content-end">
            <Button
              className={sortByScoreClass}
              onClick={onSortByScoreClick}
              color="primary"
            >
              {localization.sort.relevance}
            </Button>
            <Button
              className={sortByLastModifiedClass}
              onClick={onSortByModifiedClick}
              color="primary"
            >
              {localization.sort.modified}
            </Button>
          </div>
        </div>
      </div>
      <div className="row">
        <aside className="search-filters col-lg-4 d-none d-lg-block">
          <span className="uu-invisible" aria-hidden="false">
            Filtrering
          </span>

          <FilterPills
            history={history}
            location={location}
            locationSearch={locationSearch}
            publishers={publishers}
          />

          {informationModelAggregations && (
            <div>
              {renderFilterModal({
                showFilterModal,
                closeFilterModal,
                locationSearch,
                publisherCounts,
                publishers,
                onFilterPublisherHierarchy
              })}
              <FilterTree
                title={localization.responsible}
                aggregations={publisherCounts}
                handleFiltering={onFilterPublisherHierarchy}
                activeFilter={locationSearch.orgPath}
                referenceDataItems={publishers}
              />
            </div>
          )}
        </aside>
        <div id="informationModels" className="col-12 col-lg-8">
          {renderHits(informationModelItems, publishers)}
          <div className="col-12 d-flex justify-content-center">
            <span className="uu-invisible" aria-hidden="false">
              Sidepaginering.
            </span>
            <ReactPaginate
              pageCount={pageCount}
              pageRangeDisplayed={2}
              marginPagesDisplayed={1}
              previousLabel={localization.page.prev}
              nextLabel={localization.page.next}
              breakLabel={<span>...</span>}
              breakClassName="break-me"
              containerClassName="pagination"
              onPageChange={onPageChange}
              subContainerClassName="pages pagination"
              activeClassName="active"
              forcePage={page}
              disableInitialCallback
            />
          </div>
        </div>
      </div>
    </main>
  );
};

ResultsInformationModelPure.defaultProps = {
  showFilterModal: false,
  closeFilterModal: _.noop,

  informationModelItems: [],
  informationModelTotal: 0,
  informationModelAggregations: null,

  onFilterPublisherHierarchy: _.noop,

  publisherCounts: [],
  publishers: null,

  hitsPerPage: 10,

  history: { push: _.noop },
  location: { search: '' }
};

ResultsInformationModelPure.propTypes = {
  showFilterModal: PropTypes.bool,
  closeFilterModal: PropTypes.func,

  informationModelItems: PropTypes.array,
  informationModelTotal: PropTypes.number,
  informationModelAggregations: PropTypes.object,

  onFilterPublisherHierarchy: PropTypes.func,
  publisherCounts: PropTypes.array,
  publishers: PropTypes.object,

  hitsPerPage: PropTypes.number,

  history: PropTypes.object,
  location: PropTypes.object
};

export const ResultsInformationModel = withRouter(ResultsInformationModelPure);
