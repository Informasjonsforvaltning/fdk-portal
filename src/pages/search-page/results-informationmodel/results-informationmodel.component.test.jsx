import React from 'react';
import { shallow } from 'enzyme';
import _ from 'lodash';
import { ResultsInformationModelPure } from './results-informationmodel.component';
import informationModelApiResponse from '../__fixtures/informationmodelsApiResponse.json';

test('should render ResultsInformationModel correctly with minimum of props and no hits', () => {
  const result = shallow(<ResultsInformationModelPure />);
  expect(result).toMatchSnapshot();
});

test('should render ResultsInformationModel correctly with hits', () => {
  const props = {
    informationModelItems: _.get(informationModelApiResponse, 'hits'),
    informationModelAggregations: _.get(
      informationModelApiResponse,
      'aggregations'
    ),
    informationModelTotal: _.get(informationModelApiResponse, 'total')
  };
  const result = shallow(<ResultsInformationModelPure {...props} />);
  expect(result).toMatchSnapshot();
});
