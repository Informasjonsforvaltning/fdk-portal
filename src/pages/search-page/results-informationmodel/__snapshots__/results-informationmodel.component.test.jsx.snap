// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ResultsInformationModel correctly with hits 1`] = `
<main
  data-test-id="informationModels"
  id="content"
>
  <div
    className="row mb-3"
  >
    <div
      className="col-6 col-lg-4"
    />
    <div
      className="col-6 col-lg-4 offset-lg-4"
    >
      <div
        className="d-flex justify-content-end"
      >
        <Button
          className="fdk-button fdk-button-black-toggle selected"
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Relevance
        </Button>
        <Button
          className="fdk-button fdk-button-black-toggle"
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Recently published
        </Button>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <aside
      className="search-filters col-lg-4 d-none d-lg-block"
    >
      <span
        aria-hidden="false"
        className="uu-invisible"
      >
        Filtrering
      </span>
      <FilterPills
        history={
          Object {
            "push": [Function],
          }
        }
        location={
          Object {
            "search": "",
          }
        }
        locationSearch={Object {}}
        losItems={null}
        publishers={null}
        themesItems={null}
      />
      <div>
        <Modal
          autoFocus={true}
          backdrop={true}
          backdropTransition={
            Object {
              "mountOnEnter": true,
              "timeout": 150,
            }
          }
          centered={false}
          fade={true}
          isOpen={false}
          keyboard={true}
          modalTransition={
            Object {
              "timeout": 300,
            }
          }
          onClosed={[Function]}
          onOpened={[Function]}
          role="dialog"
          toggle={[Function]}
          zIndex={1050}
        >
          <ModalHeader
            charCode={215}
            closeAriaLabel="Close"
            tag="h5"
            toggle={[Function]}
            wrapTag="div"
          >
            Filter
          </ModalHeader>
          <ModalBody
            tag="div"
          >
            <div
              className="search-filters"
            >
              <FilterTree
                activeFilter={null}
                aggregations={Array []}
                collapseItems={false}
                handleFiltering={[Function]}
                referenceDataItems={null}
                title="Responsible"
              />
            </div>
          </ModalBody>
          <ModalFooter
            tag="div"
          >
            <Button
              className="fdk-button"
              color="primary"
              onClick={[Function]}
              tag="button"
            >
              Close
            </Button>
          </ModalFooter>
        </Modal>
        <FilterTree
          activeFilter={null}
          aggregations={Array []}
          collapseItems={false}
          handleFiltering={[Function]}
          referenceDataItems={null}
          title="Responsible"
        />
      </div>
    </aside>
    <div
      className="col-12 col-lg-8"
      id="informationModels"
    >
      <SearchHitItem
        fadeInCounter={0}
        item={
          Object {
            "id": "eadda99c-781e-4900-8f54-8dd298168382",
            "publisher": Object {
              "id": "970018131",
              "name": "UTDANNINGSDIREKTORATET",
              "orgPath": "/STAT/872417842/970018131",
              "prefLabel": Object {
                "no": "Utdanningsdirektoratet",
              },
              "uri": "http://data.brreg.no/enhetsregisteret/enhet/970018131",
            },
            "schema": Object {
              "$id": "http://localhost:8080/informationmodels/eadda99c-781e-4900-8f54-8dd298168382/schema",
              "definitions": Object {
                "Adresse": Object {
                  "properties": Object {
                    "adresse": Object {
                      "items": Object {
                        "type": "string",
                      },
                      "type": "array",
                    },
                    "kommune": Object {
                      "type": "string",
                    },
                    "kommunenummer": Object {
                      "type": "string",
                    },
                    "land": Object {
                      "type": "string",
                    },
                    "landkode": Object {
                      "type": "string",
                    },
                    "postnummer": Object {
                      "type": "string",
                    },
                    "poststed": Object {
                      "type": "string",
                    },
                  },
                },
                "Enhet": Object {
                  "properties": Object {
                    "_links": Object {
                      "$ref": "#/definitions/Self",
                    },
                    "antallAnsatte": Object {
                      "type": "integer",
                    },
                    "forretningsadresse": Object {
                      "$ref": "#/definitions/Adresse",
                    },
                    "institusjonellSektorkode": Object {
                      "$ref": "#/definitions/Sektorkode",
                    },
                    "konkurs": Object {
                      "type": "boolean",
                    },
                    "maalform": Object {
                      "type": "string",
                    },
                    "naeringskode1": Object {
                      "$ref": "#/definitions/Naeringskode",
                    },
                    "naeringskode2": Object {
                      "$ref": "#/definitions/Naeringskode",
                    },
                    "naeringskode3": Object {
                      "$ref": "#/definitions/Naeringskode",
                    },
                    "navn": Object {
                      "type": "string",
                    },
                    "organisasjonsform": Object {
                      "$ref": "#/definitions/Organisasjonsform",
                    },
                    "organisasjonsnummer": Object {
                      "type": "string",
                    },
                    "postadresse": Object {
                      "$ref": "#/definitions/Adresse",
                    },
                    "registreringsdatoEnhetsregisteret": Object {
                      "type": "string",
                    },
                    "registrertIForetaksregisteret": Object {
                      "type": "boolean",
                    },
                    "registrertIFrivillighetsregisteret": Object {
                      "type": "boolean",
                    },
                    "registrertIMvaregisteret": Object {
                      "type": "boolean",
                    },
                    "registrertIStiftelsesregisteret": Object {
                      "type": "boolean",
                    },
                    "sisteInnsendteAarsregnskap": Object {
                      "type": "string",
                    },
                    "stiftelsedato": Object {
                      "type": "string",
                    },
                    "underAvvikling": Object {
                      "type": "boolean",
                    },
                    "underTvangsavviklingEllerTvangsopplosning": Object {
                      "type": "boolean",
                    },
                  },
                },
                "Href": Object {
                  "properties": Object {
                    "href": Object {
                      "type": "string",
                    },
                  },
                },
                "Naeringskode": Object {
                  "properties": Object {
                    "beskrivelse": Object {
                      "type": "string",
                    },
                    "kode": Object {
                      "type": "string",
                    },
                  },
                },
                "Organisasjonsform": Object {
                  "properties": Object {
                    "_links": Object {
                      "$ref": "#/definitions/Self",
                    },
                    "beskrivelse": Object {
                      "type": "string",
                    },
                    "kode": Object {
                      "type": "string",
                    },
                  },
                },
                "Organisasjonsformer": Object {
                  "properties": Object {
                    "organisasjonsformer": Object {
                      "items": Object {
                        "$ref": "#/definitions/Organisasjonsform",
                      },
                      "type": "array",
                    },
                  },
                },
                "Sektorkode": Object {
                  "properties": Object {
                    "beskrivelse": Object {
                      "type": "string",
                    },
                    "kode": Object {
                      "type": "string",
                    },
                  },
                },
                "Self": Object {
                  "properties": Object {
                    "self": Object {
                      "$ref": "#/definitions/Href",
                    },
                  },
                },
                "_Organisasjonsformer": Object {
                  "properties": Object {
                    "embedded": Object {
                      "$ref": "#/definitions/Organisasjonsformer",
                    },
                  },
                },
              },
            },
            "title": "Enhetsregisteret",
            "uri": "http://localhost:8080/informationmodels/eadda99c-781e-4900-8f54-8dd298168382",
          }
        }
        key="eadda99c-781e-4900-8f54-8dd298168382"
        publishers={null}
      />
      <div
        className="col-12 d-flex justify-content-center"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Sidepaginering.
        </span>
        <PaginationBoxView
          activeClassName="active"
          breakClassName="break-me"
          breakLabel={
            <span>
              ...
            </span>
          }
          containerClassName="pagination"
          disableInitialCallback={true}
          disabledClassName="disabled"
          forcePage={0}
          marginPagesDisplayed={1}
          nextClassName="next"
          nextLabel="Neste side"
          onPageChange={[Function]}
          pageCount={1}
          pageRangeDisplayed={2}
          previousClassName="previous"
          previousLabel="Forrige side"
          subContainerClassName="pages pagination"
        />
      </div>
    </div>
  </div>
</main>
`;

exports[`should render ResultsInformationModel correctly with minimum of props and no hits 1`] = `
<main
  data-test-id="informationModels"
  id="content"
>
  <div
    className="row mb-3"
  >
    <div
      className="col-6 col-lg-4"
    />
    <div
      className="col-6 col-lg-4 offset-lg-4"
    >
      <div
        className="d-flex justify-content-end"
      >
        <Button
          className="fdk-button fdk-button-black-toggle selected"
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Relevance
        </Button>
        <Button
          className="fdk-button fdk-button-black-toggle"
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Recently published
        </Button>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <aside
      className="search-filters col-lg-4 d-none d-lg-block"
    >
      <span
        aria-hidden="false"
        className="uu-invisible"
      >
        Filtrering
      </span>
      <FilterPills
        history={
          Object {
            "push": [Function],
          }
        }
        location={
          Object {
            "search": "",
          }
        }
        locationSearch={Object {}}
        losItems={null}
        publishers={null}
        themesItems={null}
      />
    </aside>
    <div
      className="col-12 col-lg-8"
      id="informationModels"
    >
      <div
        className="col-12 d-flex justify-content-center"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Sidepaginering.
        </span>
        <PaginationBoxView
          activeClassName="active"
          breakClassName="break-me"
          breakLabel={
            <span>
              ...
            </span>
          }
          containerClassName="pagination"
          disableInitialCallback={true}
          disabledClassName="disabled"
          forcePage={0}
          marginPagesDisplayed={1}
          nextClassName="next"
          nextLabel="Neste side"
          onPageChange={[Function]}
          pageCount={1}
          pageRangeDisplayed={2}
          previousClassName="previous"
          previousLabel="Forrige side"
          subContainerClassName="pages pagination"
        />
      </div>
    </div>
  </div>
</main>
`;
